### PR TITLE
Vulkan rendering backend. Early Work-In-Progress

### PR DESCRIPTION
**Pull request by hrydgard**
Please note - this is not for general use, it's only an early preview. It's very rough, only works in non-buffered (crashes in buffered) and is missing many features and optimizations, so it's not even fast yet.

EDIT: At this point, it already beats OpenGL in nearly everything I've tried, though is not feature complete of course.

But, it's a start, and this will hopefully let us take performance to the next level on Android, eventually - once drivers supporting Vulkan are released, and present on devices. Some devices will get Vulkan with a Marshmallow update (nVidia Shield/TV) while others will have to wait longer or won't get it at all.

### CODE REVIEW

**Comment by Anuskuss** [comment]
Awesome :+1:

**Comment by GAMELASTER** [comment]
D3D11 backend WIP too?

**Comment by hrydgard** [comment]
@GAMELASTER  DX11? only vague plans, nothing concrete. Just made space in the enum, heh. Might remove it from the list.

**Comment by VIRGINKLM** [comment]
Just a small note. Samsung Galaxy A6 (2016) supports Vulkan straight out of the box and NVidia Shield devices began recieving Dev buils supporting Vulkan. Also I bet Galaxy S7 variants will have straight out of the box support for it.

**Comment by Zetchzie** [comment]
So will this work on ogl backend in the future and can this hold potential for better performance on damanding games on hardware that couldnt run at full speed previously?

**Comment by zminhquanz** [comment]
https://en.wikipedia.org/wiki/Vulkan_(API)
Here is the list for the GPU from NVIDIA , AMD and PowerVR , Mali , Adreno , Intel to support Vulkan API
Device can be update to Android 6.0 to support it

File: Common/GraphicsContext.h [hunk 1]
```diff
@@ -22,6 +22,9 @@ class GraphicsContext {
 
 	virtual void Resize() = 0;
 
+	// Needs casting to the appropriate type, unfortunately. Should find a better solution..
+	virtual void *GetAPIContext() { return nullptr; }
```
Comment (unknownbrackets) [inline Common/GraphicsContext.h]: Seems like this is no longer used?

-[Unknown]

File: GPU/Directx9/TextureCacheDX9.cpp [hunk 2]
```diff
@@ -1639,6 +1639,7 @@ TextureCacheDX9::TexCacheEntry::Status TextureCacheDX9::CheckAlpha(const u32 *pi
 	return (TexCacheEntry::Status)res;
 }
 
+// TODO: xoffset and yoffset are unused - bug?
```
Comment (unknownbrackets) [inline GPU/Directx9/TextureCacheDX9.cpp]: Hmm, they're fixed at 0 also.  Maybe the plan was to support updating a subarea?  Sounds hard, maybe we should just remove.

-[Unknown]

File: GPU/Vulkan/DrawEngineVulkan.cpp [hunk 3]
```diff
@@ -0,0 +1,733 @@
...
+	decoded = (u8 *)AllocateMemoryPages(DECODED_VERTEX_BUFFER_SIZE);
+	decIndex = (u16 *)AllocateMemoryPages(DECODED_INDEX_BUFFER_SIZE);
+	splineBuffer = (u8 *)AllocateMemoryPages(SPLINE_BUFFER_SIZE);
+	transformed = (TransformedVertex *)AllocateMemoryPages(TRANSFORMED_VERTEX_BUFFER_SIZE);
+	transformedExpanded = (TransformedVertex *)AllocateMemoryPages(3 * TRANSFORMED_VERTEX_BUFFER_SIZE);
+
+	indexGen.Setup(decIndex);
+
+	// All resources we need for PSP drawing. Usually only bindings 0 and 2-4 are populated.
+	VkDescriptorSetLayoutBinding bindings[5];
```
Comment (unknownbrackets) [inline GPU/Vulkan/DrawEngineVulkan.cpp]: Should we zero initialize in case 1.1 adds more members?

-[Unknown]

File: Windows/MainWindow.cpp [hunk 4]
```diff
@@ -701,13 +702,22 @@ namespace MainWindow
...
 			break;
 
+		case WM_ENTERSIZEMOVE:
+			inResizeMove = true;
+			break;
+
+		case WM_EXITSIZEMOVE:
+			inResizeMove = false;
+			HandleSizeChange(SIZE_RESTORED);
+			break;
```
Comment (unknownbrackets) [inline Windows/MainWindow.cpp]: Are these changes Vulkan related or separate?

-[Unknown]

File: GPU/Vulkan/DrawEngineVulkan.cpp (hunk 3 above)
Comment (hrydgard) [inline GPU/Vulkan/DrawEngineVulkan.cpp]: Vulkan is not extended that way for minor versions, instead, additional structs are chained using the pNext member, which is more ABI safe. Changing structs requires a major version upgrade. There's a whole chapter on how Vulkan will be extended in vkspec.pdf :)

But yeah, might as well zero it anyway.

File: Windows/MainWindow.cpp (hunk 4 above)
Comment (hrydgard) [inline Windows/MainWindow.cpp]: Commited separately to master.

**Comment by unknownbrackets** [comment]
Where is the graphicsContext supposed to be populated?  I had to add `PSP_CoreParameter().graphicsContext = graphicsContext;` to `NativeRender()` to fix some crashes.

Well, now I'm crashing in vkCmdClearAttachments, but it's progress.

-[Unknown]

**Comment by hrydgard** [comment]
@unknownbrackets Remember to turn off buffered rendering, it will crash :)

Don't recall having problems with graphicsContext... hm

**Comment by hrydgard** [comment]
@unknownbrackets  turns out I broke depth buffer init, too, when messing with the shield

File: Windows/PPSSPP.sln [hunk 5]
```diff
@@ -75,18 +76,30 @@ Project("{8BC9CEB8-8B4A-11D0-8D11-00A0C91BC942}") = "glslang", "..\ext\glslang.v
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug CRT DLL|Win32 = Debug CRT DLL|Win32
+		Debug CRT DLL|x64 = Debug CRT DLL|x64
 		Debug|Win32 = Debug|Win32
 		Debug|x64 = Debug|x64
+		Release CRT DLL|Win32 = Release CRT DLL|Win32
+		Release CRT DLL|x64 = Release CRT DLL|x64
```
Comment (unknownbrackets) [inline Windows/PPSSPP.sln]: Do we want to add these, or were they just added by something else?

-[Unknown]

File: Windows/WindowsHost.cpp [hunk 6]
```diff
@@ -181,6 +187,7 @@ void WindowsHost::SetDebugMode(bool mode) {
 }
 
 void WindowsHost::PollControllers(InputState &input_state) {
+	return;
```
Comment (unknownbrackets) [inline Windows/WindowsHost.cpp]: Temp hack warning.

-[Unknown]

Comment (hrydgard) [inline Windows/WindowsHost.cpp]: oops indeed

File: Windows/PPSSPP.sln (hunk 5 above)
Comment (hrydgard) [inline Windows/PPSSPP.sln]: Hm, they were probably added when I tried to import the CMake-generated project for glslang, before I made my own instead.

File: unittest/UnitTests.vcxproj [hunk 7]
```diff
@@ -123,7 +122,7 @@
     <Link>
       <SubSystem>Console</SubSystem>
       <GenerateDebugInformation>true</GenerateDebugInformation>
-      <AdditionalDependencies>Ws2_32.lib;winmm.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;opengl32.lib;dsound.lib;glu32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;comctl32.lib;d3d9.lib;dxguid.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>..\Windows\x64\Debug\glslang.lib;..\ext\vulkan\vulkan-1.lib;Ws2_32.lib;winmm.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;opengl32.lib;dsound.lib;glu32.lib;avcodec.lib;avformat.lib;avutil.lib;swresample.lib;swscale.lib;comctl32.lib;d3d9.lib;dxguid.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (unknownbrackets) [inline unittest/UnitTests.vcxproj]: Any reason only this one gets the glslang lib?

-[Unknown]

Comment (hrydgard) [inline unittest/UnitTests.vcxproj]: It's not supposed to be there, it's linked through "References". this was an old thing that slipped through some cleanup.

**Comment by unknownbrackets** [comment]
@hrydgard
> Refactor away nextTexture_. Not needed on Vulkan.

Do you mean because you plan to remove the hardware transform -> software fallback and move texture decoding earlier so that it can be done before `SetTexture()`?  Because that is the reason `nextTexture_` exists.  We don't really need it for D3D9 either, technically, I suppose.

To clarify, `ApplyTexture()` MUST be called AFTER the vertices have been decoded.  I'm not sure, but right now this isn't happening, and this could be contributing to rendering bugs.  At the very least, it's ignoring the output of the vertex decode (texture bounds.)

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, I intend to move things around a bit, I've just been swamped. And yes, UV scale/offset seems a bit broken, among other things, as expected...

File: Core/Config.h [hunk 8]
```diff
@@ -50,10 +50,14 @@ enum BufferFilter {
 enum class GPUBackend {
 	OPENGL = 0,
 	DIRECT3D9 = 1,
+  DIRECT3D11 = 2,
+  VULKAN = 3,
```
Comment (unknownbrackets) [inline Core/Config.h]: Indent nitpick.

-[Unknown]

**Comment by unknownbrackets** [comment]
Personally, I would prefer this convention:

```c++
  VkFramebufferCreateInfo fb_info = { VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO };
  fb_info.renderPass = surface_render_pass_;
```

Instead of:
```c++
  VkFramebufferCreateInfo fb_info = {};
  fb_info.sType = VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO;
  fb_info.pNext = NULL;
  fb_info.renderPass = surface_render_pass_;
```

What do you think?

-[Unknown]

**Comment by hrydgard** [comment]
Hm, not bad. I guess the compiler will eliminate the extra stores (first zeroing the struct, then filling it out) anyway. Let's do that.

**Comment by unknownbrackets** [comment]
Do you think it would be a good idea to decode directly onto the push buffer (from the jit), and then if we're caching, issue a cmd to copy that to the cache buffer (which would then be used for subsequent draws?)

By the way, I'm seeing something funny with vertex coords in FF4, seems to be why only the first letter is drawing in some places.  Still debugging, but this area has me thinking about vertex decode is all.

-[Unknown]

File: GPU/Vulkan/ShaderManagerVulkan.cpp [hunk 9]
```diff
@@ -162,7 +162,7 @@ static void ConvertProjMatrixToVulkan(Matrix4x4 &in, bool invertedX, bool invert
 	if (invertedY)
 		yoff = -yoff;
 
-	const Vec3 trans(xoff, yoff, gstate_c.vpZOffset + 0.5f);
+	const Vec3 trans(xoff, yoff, gstate_c.vpZOffset * 0.5f + 0.5f);
```
Comment (unknownbrackets) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: Wait wait wait, does Vulkan use the half-pixel offset insanity that Direct3D9 does?  It doesn't right?

-[Unknown]

-[Unknown]

File: GPU/Vulkan/ShaderManagerVulkan.cpp [hunk 10]
```diff
@@ -0,0 +1,581 @@
...
+	// Half pixel offset hack
+	float xoff = 0.5f / gstate_c.curRTRenderWidth;
+	xoff = gstate_c.vpXOffset + (invertedX ? xoff : -xoff);
+	float yoff = 0.5f / gstate_c.curRTRenderHeight;
+	yoff = gstate_c.vpYOffset + (invertedY ? yoff : -yoff);
+
+	if (invertedX)
+		xoff = -xoff;
+	if (invertedY)
+		yoff = -yoff;
```
Comment (unknownbrackets) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: I don't remember properly why we do 5667f348055c59eaee51a2498d0a0b1c3de3be26, but I don't think we need/want to do it in Vulkan?  This is what causes FF4 to not render some stuff.

-[Unknown]

Comment (hrydgard) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: Vulkan does not need the half-pixel offset, should remove that indeed.

Vulkan supports very large viewports on all hardware I've seen limits for on gpuinfo.org so we should be able to get rid of the viewport clamping stuff, and in that case the offset should be 0 and this shouldn't matter.

Are you saying FF4 renders fully if you remove these lines?

Comment (unknownbrackets) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: Yes, I didn't find anything else missing with those lines removed.  They are causing it to render some text off screen.  We don't do this in GLES.

Also, the offset will always be the drift from left to right, so it ought to always be positive, no?  I'm trying to understand why this would be correct on Direct3D 9.  Even if the coordinate system is negated, if the center is toward the right, the final coordinates should be translated positive...

-[Unknown]

Comment (hrydgard) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: Can't figure it out either, I think it was only meant to negate the half-pixel offset, not the clipping offset...

Comment (unknownbrackets) [inline GPU/Vulkan/ShaderManagerVulkan.cpp]: Hmm.  Now that I think harder about it, should the half-pixel offset be negated either?  It is also a drift in the final mapping of the pixel to screen coordinates.  It ought to have the same properties...

Also, it's being negated twice in vulkan/d3d9...

-[Unknown]

**Comment by unknownbrackets** [comment]
It's rendering now on my Android TV, although rotated 180 degrees.  I'm not sure what fixed it, do you know?

-[Unknown]

**Comment by hrydgard** [comment]
Yeah. Flipped vertically sounds bad though...

**Comment by hrydgard** [comment]
Ah, the upside down thing (an upside down mirror, not 180 degrees) is just that part of the emulator thinks it's in Vulkan when it's not, so it doesn't flip the projection matrix.

**Comment by unknownbrackets** [comment]
Yeah, sorry, I corrected my comment right after I made it - at first I assumed it was 180 degrees because of some input problems, then when I looked again I realized it was just mirrored (I'm not sure about the input problem but I can no longer reproduce, maybe I was just tired... it wasn't accepting most arrows and was doing odd things.)

That's also when I realized it must just be using GL just as you say, and edited my comment for that as well.

-[Unknown]
